### PR TITLE
cmd_

### LINKED ISSUE #234: cmd ergonomics: making own cmd derivations
Currently it's not easy to define a derivation of `cmd` without losing its special property -- vararg-ness.
And indeed, some derivations don't have an obvious way of making them.

Exhibit one: ..to avoid specifying the `unit $` prefix before each cmd invocation:
```haskell
type a :-> t = a
cmdn ∷ S.CmdArguments args => args :-> Action r
cmdn = unit . S.cmd
```
..fails type checking.

### PR DESCRIPTION
**Pull request by mikeplus64**
Fixes #234

### CODE REVIEW

File: src/Development/Shake/Command.hs [hunk 1]
```diff
@@ -1,4 +1,4 @@
-{-# LANGUAGE FlexibleInstances, TypeSynonymInstances, TypeOperators, ScopedTypeVariables, NamedFieldPuns #-}
+{-# LANGUAGE FlexibleInstances, GADTs, TypeSynonymInstances, TypeOperators, ScopedTypeVariables, MultiParamTypeClasses, NamedFieldPuns #-}
```
Comment (ndmitchell) [inline src/Development/Shake/Command.hs]: I don't see a use of either GADTs or MPTC in the code below - are they necessary?

File: src/Development/Shake/Command.hs [hunk 2]
```diff
@@ -561,6 +561,15 @@ type a :-> t = a
 cmd :: CmdArguments args => args :-> Action r
 cmd = cmdArguments []
 
+class Unit a
+instance {-# OVERLAPPING #-} Unit b => Unit (a -> b)
+instance {-# OVERLAPPABLE #-} a ~ () => Unit (m a)
```
Comment (ndmitchell) [inline src/Development/Shake/Command.hs]: The `OVERLAPPING` and `OVERLAPPABLE` are not available until GHC 7.10, and Shake still supports back to GHC 7.4. I think you can tweak it to be a language pragma on those versions - may require CPP though.

**Comment by ndmitchell** [comment]
Awesome! I didn't know this was possible, A few things to clean up, but a great addition!

File: src/Development/Shake/Command.hs (hunk 1 above)
Comment (mikeplus64) [inline src/Development/Shake/Command.hs]: MPTC no, whoops. GADTs (or TypeFamilies and probably a few other extensions) allow the `a ~ ()` constraint.

File: src/Development/Shake/Command.hs (hunk 2 above)
Comment (mikeplus64) [inline src/Development/Shake/Command.hs]: Drat -- you're right. New commits split Unit into its own module that imports yet another module based on which GHC version is available. [apparently GHC doesn't use cpp to parse language pragmas so something like
```
{-# LANGUAGE CPP #-}
#if foo
{-# LANGUAGE OverlappingInstances
#endif
...
```
didn't work]

**Comment by ndmitchell** [comment]
Awesome - looks great. I'll need to play with it a bit this weekend to check it really is matching my expectation, but a fantastic addition. Shame GHC makes it so hard!

**Comment by mikeplus64** [comment]
I tried again with just having Unit defined in the one module and it worked -- maybe I made a typo or something earlier. But I haven't tested it with GHC 7.{4,8}, hopefully the CI can illuminate any problems ...

**Comment by ndmitchell** [comment]
All works nicely - many thanks!

With your revised code I had a shot at merging the Unit module into Command and it worked, so I guess whatever you managed to fix worked there too.

**Comment by mikeplus64** [comment]
Yay!

I don't feel particularly strongly on it, but I think that having a separate module for Unit is better so that *all* the typeclass instances in Command.hs aren't marked as overlapping in GHC <7.10. I can only imagine it being a problem if an instance head in Command.hs is written incorrectly, though it still does seem like Better DesignⓇ to me.

**Comment by ndmitchell** [comment]
Good point. I'm not so worried about that since GHC 7.8 is mostly legacy by now, and if they did overlap, GHC 8.0 would catch them anyway. If it had to be that way in newer compiler versions I'd have gone the separate module route.

**Pull request merged**